moebot_bot/bot/commands/timer.go

```diff
@@ -7,46 +7,132 @@ import (
 		RWMutex: sync.RWMutex{},
-		M:       make(map[string]time.Time),
+		M:       make(map[string]*ChannelTimer),
 	}
 	return tc
 }
 
 func (tc *TimerCommand) Execute(pack *CommPackage) {
 	channelID := pack.message.ChannelID
 	if len(pack.params) > 0 && strings.EqualFold(pack.params[0], "start") {
+		// Make sure the user has at least mod-level permissions before starting the timer
 		if tc.Checker.HasPermission(pack.message.Author.ID, pack.member.Roles, pack.guild, db.PermMod) {
 			tc.chTimers.Lock()
-			tc.chTimers.M[channelID] = time.Now()
+
+			// If this channel timer is currently writing out, tell it to stop
+			if chTimer, ok := tc.chTimers.M[channelID]; ok {
+				chTimer.Lock()
+				if chTimer.isWriting {
+					close(chTimer.requestCh)
+				}
+				chTimer.Unlock()
+			}
+
+			// Create a new timer
+			tc.chTimers.M[channelID] = &ChannelTimer{
+				time:      time.Now(),
+				writes:    0,
+				isWriting: false,
+				requestCh: make(chan string, 10),
+			}
+
 			tc.chTimers.Unlock()
 			pack.session.ChannelMessageSend(pack.message.ChannelID, "Timer started!")
 		} else {
 			pack.session.ChannelMessageSend(pack.message.ChannelID, pack.message.Author.Mention()+", you... you don't have permission to do that!")
 		}
 	} else {
 		tc.chTimers.RLock()
-		if v, ok := tc.chTimers.M[channelID]; ok {
-			pack.session.ChannelMessageSend(pack.message.ChannelID, fmtDuration(time.Since(v)))
+		if chTimer, ok := tc.chTimers.M[channelID]; ok {
+			chTimer.Lock()
+			// Reset the number of writes
+			chTimer.writes = 0
+
+			// If the time is not writing, start it
+			if !chTimer.isWriting {
+				go chTimer.writeTimes(pack)
+				chTimer.isWriting = true
+			}
+			chTimer.Unlock()
 		} else {
 			pack.session.ChannelMessageSend(pack.message.ChannelID, "No timer started for this channel...")
 		}
 		tc.chTimers.RUnlock()
 	}
 }
 
+func (ct *ChannelTimer) writeTimes(pack *CommPackage) {
+	duration := time.Since(ct.time)
+
+	// Write the time once right away
+	go func() {
+		pack.session.ChannelMessageSend(pack.message.ChannelID, fmtDuration(duration))
+		ct.Lock()
+		ct.writes++
+		ct.Unlock()
+	}()
+
+	// Synchronize the writes to be divisible by the interval (works well when interval is 5 so we get writes at times like 0:30, 0:35, 0:40, etc.)
+	timeToSync := writeInterval - (duration % writeInterval)
+	time.Sleep(timeToSync)
+	duration += timeToSync
+
+	// Write again if we spent a sufficient time syncing, otherwise just wait until the next write interval
+	go func() {
+		if timeToSync > time.Second {
+			ct.Lock()
+			pack.session.ChannelMessageSend(pack.message.ChannelID, fmtDuration(duration))
+			ct.writes++
+			ct.Unlock()
+		}
+	}()
+
+	// Start writing until we reach the max number of writes or get a message to stop
+	for {
+		select {
+		case msg, chOpen := <-ct.requestCh:
+			// Break out of this loop if the channel was closed (probably shouldn't be closed since there are multiple writeres) or a "stop" was issued
+			if !chOpen || msg == "stop" {
```

Comment:
Since we don't actually ever pass in `stop` guessing this is left in for future updates? (I'm okay with it)

Reply:
Rather an oversight on my end to leave it in. I was struggling internally with an awkward situation of only using a channel to check for a close vs. checking the message sent for only "stop" commands which should only come once.

I'm going to remove the checking for "stop" to make it more clear what the channel is expecting.